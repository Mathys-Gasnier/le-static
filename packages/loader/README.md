# loader

A library used to load project folder structure into a tree used by other packages