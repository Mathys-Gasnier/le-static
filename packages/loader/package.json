{
  "name": "@le-static/loader",
  "version": "1.0.1",
  "license": "Apache-2.0",
  "author": "Mathys Gasnier",
  "homepage": "https://github.com/Mathys-Gasnier/le-static/tree/master/packages/loader",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Mathys-Gasnier/le-static.git"
  },
  "bugs": {
    "url": "https://github.com/Mathys-Gasnier/le-static/issues"
  },
  "dependencies": {
    "tslib": "^2.3.0"
  },
  "type": "commonjs",
  "main": "./src/index.js",
  "typings": "./src/index.d.ts"
}