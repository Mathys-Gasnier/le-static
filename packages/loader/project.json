{
  "name": "loader",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/loader/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nx/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/loader",
        "main": "packages/loader/src/index.ts",
        "tsConfig": "packages/loader/tsconfig.lib.json",
        "assets": ["packages/loader/*.md"]
      }
    },
    "publish": {
      "command": "node tools/scripts/publish.mjs loader {args.ver} {args.tag}",
      "dependsOn": ["build"]
    },
    "lint": {
      "executor": "@nx/eslint:lint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": [
          "packages/loader/**/*.ts",
          "packages/loader/package.json"
        ]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "packages/loader/jest.config.ts"
      }
    }
  },
  "tags": []
}
