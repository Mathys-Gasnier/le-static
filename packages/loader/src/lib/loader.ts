import { existsSync, readFileSync, readdirSync } from 'fs';

export interface File {
  type: 'file',
  name: string,
  extention: string,
  content: string
}

export interface Folder {
  type: 'folder',
  files: {
    [key: string]: File | Folder
  }
}

export interface Project {
  pages: Folder,
  components: Folder,
  resources: Folder,
  styles: Folder
}

export function load(src: string): Project {

  const project: Project = {
    pages: loadFolder(`${src}/pages`),
    components: loadFolder(`${src}/components`),
    resources: loadFolder(`${src}/resources`),
    styles: loadFolder(`${src}/styles`)
  }

  return project;
}

function loadFolder(src: string): Folder {
  const folder: Folder = { type: 'folder', files: { } };

  if(!existsSync(src)) return folder;

  for(const file of readdirSync(src, { withFileTypes: true })) {
    if(file.isDirectory()) folder.files[file.name] = loadFolder(`${src}/${file.name}`);
    else {
      const name = file.name.split('.').slice(0, -1).join('.');
      const extention = file.name.split('.').slice(-1).join('');
      const content = readFileSync(`${src}/${file.name}`);

      folder.files[file.name] = {
        type: 'file',
        name: name,
        extention: extention,
        content: content.toString()
      }
    }
  }

  return folder;
}