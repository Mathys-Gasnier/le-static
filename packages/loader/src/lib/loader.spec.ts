import { load } from './loader';

describe('loader', () => {
  it('should work', () => {
    expect(load('C:\\Users\\matbo\\Documents\\le-static\\test\\simple')).toEqual({
      pages: {
        'index.md': {
          name: 'index',
          extention: 'md',
          content: '<<< file.css\r\n' +
            '\r\n' +
            '# Title\r\n' +
            '\r\n' +
            '## Sub Title\r\n' +
            '\r\n' +
            '### SUB SUB Title\r\n' +
            '\r\n' +
            'this\r\n' +
            'is a \r\n' +
            'paragraph\r\n' +
            '\r\n' +
            '---\r\n' +
            '\r\n' +
            'another\r\n' +
            'paragraph\r\n' +
            '\r\n' +
            '<<< cool/footer.md'
        }
      },
      components: { cool: { 'footer.md': {
        content: '## Footer\r\n' +
          '\r\n' +
          'This is a cool footer',
        extention: 'md',
        name: 'footer'
      } } },
      resources: {},
      styles: {
        'file.css': {
          name: 'file',
          extention: 'css',
          content: '\r\n.paragraph {\r\n    color: blue;\r\n}'
        },
        'index.css': {
          name: 'index',
          extention: 'css',
          content: '\r\n.separator {\r\n    margin: 5vw;\r\n}'
        }
      }
    });
  });
});
