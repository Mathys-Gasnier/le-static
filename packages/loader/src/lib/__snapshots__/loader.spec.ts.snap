// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader Should properly load project folders 1`] = `
{
  "components": {
    "files": {
      "cool": {
        "files": {
          "footer.md": {
            "content": {
              "data": [
                35,
                35,
                32,
                70,
                111,
                111,
                116,
                101,
                114,
                10,
                10,
                84,
                104,
                105,
                115,
                32,
                105,
                115,
                32,
                97,
                32,
                99,
                111,
                111,
                108,
                32,
                102,
                111,
                111,
                116,
                101,
                114,
              ],
              "type": "Buffer",
            },
            "extention": "md",
            "name": "footer",
            "type": "file",
          },
        },
        "type": "folder",
      },
    },
    "type": "folder",
  },
  "config": {
    "build": {
      "components": {
        "suffix": "cool/footer.md",
      },
      "outDir": "dist",
    },
    "server": {
      "port": 5454,
    },
    "site": {
      "title": "Demo - %page_name%",
    },
  },
  "pages": {
    "files": {
      "folder": {
        "files": {
          "index.md": {
            "content": {
              "data": [
                64,
                32,
                70,
                111,
                108,
                100,
                101,
                114,
                10,
                10,
                35,
                32,
                70,
                111,
                108,
                100,
                101,
                114,
                10,
                10,
                91,
                66,
                97,
                99,
                107,
                32,
                116,
                111,
                32,
                109,
                97,
                105,
                110,
                93,
                40,
                47,
                41,
              ],
              "type": "Buffer",
            },
            "extention": "md",
            "name": "index",
            "type": "file",
          },
        },
        "type": "folder",
      },
      "index.md": {
        "content": {
          "data": [
            64,
            32,
            77,
            97,
            105,
            110,
            32,
            80,
            97,
            103,
            101,
            10,
            64,
            117,
            115,
            101,
            32,
            102,
            105,
            108,
            101,
            46,
            99,
            115,
            115,
            10,
            10,
            35,
            32,
            77,
            97,
            105,
            110,
            32,
            84,
            105,
            116,
            108,
            101,
            10,
            10,
            35,
            35,
            32,
            83,
            117,
            98,
            32,
            84,
            105,
            116,
            108,
            101,
            10,
            10,
            35,
            35,
            35,
            32,
            83,
            85,
            66,
            32,
            83,
            85,
            66,
            32,
            84,
            105,
            116,
            108,
            101,
            10,
            10,
            116,
            104,
            105,
            115,
            10,
            105,
            115,
            32,
            97,
            32,
            10,
            112,
            97,
            114,
            97,
            103,
            114,
            97,
            112,
            104,
            10,
            10,
            45,
            45,
            45,
            10,
            10,
            97,
            110,
            111,
            116,
            104,
            101,
            114,
            10,
            112,
            97,
            114,
            97,
            103,
            114,
            97,
            112,
            104,
            10,
            10,
            100,
            105,
            102,
            102,
            101,
            114,
            101,
            110,
            116,
            32,
            116,
            121,
            112,
            101,
            32,
            111,
            102,
            32,
            116,
            101,
            120,
            116,
            58,
            10,
            45,
            32,
            42,
            105,
            116,
            97,
            108,
            105,
            99,
            42,
            10,
            45,
            32,
            42,
            42,
            98,
            111,
            108,
            100,
            42,
            42,
            10,
            45,
            32,
            42,
            42,
            42,
            66,
            111,
            108,
            100,
            32,
            73,
            116,
            97,
            108,
            105,
            99,
            42,
            42,
            42,
            10,
            45,
            32,
            96,
            67,
            111,
            100,
            101,
            32,
            119,
            111,
            114,
            100,
            115,
            96,
            10,
            10,
            49,
            46,
            32,
            111,
            110,
            101,
            10,
            50,
            46,
            32,
            116,
            119,
            111,
            10,
            10,
            62,
            32,
            84,
            104,
            105,
            115,
            32,
            105,
            115,
            32,
            115,
            111,
            109,
            101,
            32,
            105,
            109,
            112,
            111,
            114,
            116,
            97,
            110,
            116,
            10,
            62,
            32,
            115,
            116,
            117,
            102,
            102,
            32,
            115,
            97,
            105,
            100,
            32,
            98,
            121,
            32,
            115,
            109,
            97,
            114,
            116,
            32,
            112,
            101,
            111,
            112,
            108,
            101,
            10,
            62,
            32,
            73,
            110,
            32,
            116,
            104,
            111,
            115,
            101,
            32,
            98,
            111,
            120,
            101,
            115,
            10,
            10,
            96,
            96,
            96,
            106,
            115,
            10,
            99,
            111,
            110,
            115,
            111,
            108,
            101,
            46,
            108,
            111,
            103,
            40,
            39,
            116,
            101,
            115,
            116,
            39,
            41,
            59,
            10,
            96,
            96,
            96,
            10,
            10,
            64,
            105,
            109,
            112,
            111,
            114,
            116,
            32,
            99,
            111,
            111,
            108,
            47,
            102,
            111,
            111,
            116,
            101,
            114,
            46,
            109,
            100,
            10,
            10,
            33,
            91,
            66,
            114,
            111,
            119,
            110,
            32,
            83,
            113,
            117,
            97,
            114,
            101,
            93,
            40,
            66,
            114,
            111,
            119,
            110,
            46,
            112,
            110,
            103,
            41,
            10,
            10,
            91,
            116,
            111,
            32,
            70,
            111,
            108,
            100,
            101,
            114,
            32,
            49,
            93,
            40,
            47,
            102,
            111,
            108,
            100,
            101,
            114,
            47,
            105,
            110,
            100,
            101,
            120,
            46,
            109,
            100,
            41,
          ],
          "type": "Buffer",
        },
        "extention": "md",
        "name": "index",
        "type": "file",
      },
    },
    "type": "folder",
  },
  "resources": {
    "files": {
      "Brown.png": {
        "content": {
          "data": [
            137,
            80,
            78,
            71,
            13,
            10,
            26,
            10,
            0,
            0,
            0,
            13,
            73,
            72,
            68,
            82,
            0,
            0,
            0,
            128,
            0,
            0,
            0,
            128,
            8,
            6,
            0,
            0,
            0,
            195,
            62,
            97,
            203,
            0,
            0,
            0,
            1,
            115,
            82,
            71,
            66,
            0,
            174,
            206,
            28,
            233,
            0,
            0,
            0,
            4,
            103,
            65,
            77,
            65,
            0,
            0,
            177,
            143,
            11,
            252,
            97,
            5,
            0,
            0,
            0,
            9,
            112,
            72,
            89,
            115,
            0,
            0,
            14,
            194,
            0,
            0,
            14,
            194,
            1,
            21,
            40,
            74,
            128,
            0,
            0,
            1,
            80,
            73,
            68,
            65,
            84,
            120,
            94,
            237,
            210,
            49,
            1,
            192,
            32,
            16,
            192,
            192,
            167,
            150,
            16,
            136,
            116,
            186,
            224,
            34,
            119,
            75,
            12,
            100,
            157,
            61,
            119,
            200,
            250,
            94,
            137,
            50,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            206,
            0,
            113,
            6,
            136,
            51,
            64,
            156,
            1,
            226,
            12,
            16,
            103,
            128,
            56,
            3,
            196,
            25,
            32,
            109,
            230,
            7,
            53,
            28,
            2,
            177,
            183,
            83,
            113,
            238,
            0,
            0,
            0,
            0,
            73,
            69,
            78,
            68,
            174,
            66,
            96,
            130,
          ],
          "type": "Buffer",
        },
        "extention": "png",
        "name": "Brown",
        "type": "file",
      },
    },
    "type": "folder",
  },
  "src": "./test/simple",
  "styles": {
    "files": {
      "file.css": {
        "content": {
          "data": [
            10,
            46,
            112,
            97,
            114,
            97,
            103,
            114,
            97,
            112,
            104,
            32,
            123,
            10,
            32,
            32,
            32,
            32,
            99,
            111,
            108,
            111,
            114,
            58,
            32,
            98,
            108,
            117,
            101,
            59,
            10,
            125,
          ],
          "type": "Buffer",
        },
        "extention": "css",
        "name": "file",
        "type": "file",
      },
      "index.css": {
        "content": {
          "data": [
            10,
            112,
            32,
            123,
            10,
            32,
            32,
            32,
            32,
            119,
            104,
            105,
            116,
            101,
            45,
            115,
            112,
            97,
            99,
            101,
            58,
            32,
            112,
            114,
            101,
            59,
            10,
            125,
            10,
            10,
            46,
            115,
            101,
            112,
            97,
            114,
            97,
            116,
            111,
            114,
            32,
            123,
            10,
            32,
            32,
            32,
            32,
            100,
            105,
            115,
            112,
            108,
            97,
            121,
            58,
            32,
            98,
            108,
            111,
            99,
            107,
            59,
            10,
            32,
            32,
            32,
            32,
            119,
            105,
            100,
            116,
            104,
            58,
            32,
            49,
            48,
            48,
            37,
            59,
            10,
            32,
            32,
            32,
            32,
            104,
            101,
            105,
            103,
            104,
            116,
            58,
            32,
            49,
            112,
            120,
            59,
            10,
            32,
            32,
            32,
            32,
            98,
            111,
            114,
            100,
            101,
            114,
            45,
            116,
            111,
            112,
            58,
            32,
            49,
            112,
            120,
            32,
            115,
            111,
            108,
            105,
            100,
            32,
            98,
            108,
            97,
            99,
            107,
            59,
            10,
            125,
            10,
            10,
            46,
            98,
            108,
            111,
            99,
            107,
            45,
            113,
            117,
            111,
            116,
            101,
            115,
            32,
            123,
            10,
            32,
            32,
            32,
            32,
            109,
            97,
            114,
            103,
            105,
            110,
            45,
            108,
            101,
            102,
            116,
            58,
            32,
            51,
            101,
            109,
            59,
            10,
            125,
            10,
            10,
            46,
            99,
            111,
            100,
            101,
            98,
            108,
            111,
            99,
            107,
            32,
            123,
            10,
            32,
            32,
            32,
            32,
            98,
            97,
            99,
            107,
            103,
            114,
            111,
            117,
            110,
            100,
            45,
            99,
            111,
            108,
            111,
            114,
            58,
            32,
            35,
            99,
            99,
            99,
            99,
            99,
            99,
            59,
            10,
            32,
            32,
            32,
            32,
            112,
            97,
            100,
            100,
            105,
            110,
            103,
            58,
            32,
            49,
            101,
            109,
            59,
            10,
            125,
          ],
          "type": "Buffer",
        },
        "extention": "css",
        "name": "index",
        "type": "file",
      },
    },
    "type": "folder",
  },
}
`;
